### ISSUE TITLE
Parsing of responses with empty body and no Content-Type

### ISSUE THREAD

**Issue by radarsh** [http-builder-ng/http-builder-ng#149]
I don't know yet if this is an issue or simply a gap in my understanding of how the library works but please hear me out first.

## TL; DR
I have trouble parsing responses with no body and get a `java.lang.ClassCastException: Cannot cast [B to com.foo.ApiResponse`.

## Background

I have a scenario where an endpoint can return a `201` response with empty body and no `Content-Type` header. In my code I have an abstraction over the http builder so that I don't expose http builder API to the callers of this abstraction. Let's call this abstraction `ApiService`. The general contract of `ApiService` methods is something like this:

### ApiRequest (partial code)
```groovy
class ApiRequest {
    String path
    String contentType
    Map<String, Object> query
    Map<String, String> headers
    Map body
}
```

### ApiResponse (partial code)
```groovy
class ApiResponse {
    int status
    Map body
    List bodyList
    List<HttpCookie> cookies
    List<Map<String, String>> headers
}
```

### ApiService (partial code)
```groovy
ApiResponse get(ApiRequest)
ApiResponse post(ApiRequest)
ApiResponse put(ApiRequest)
```

In `ApiService` I configure an instance of `HttpBuilder` and make sure every type of response is eventually converted into an instance of `ApiResponse`. To do this, I have response parsers defined.

```groovy
response.parser(MEDIA_TYPES) { ChainedHttpConfig config, FromServer fs ->
    new ApiResponse(fs)
}

// I don't really want to add a 'when' for every single status that can return empty body
response.when(CREATED) { FromServer fs ->
    new ApiResponse(fs)
}

response.when(NO_CONTENT.value()) { FromServer fs ->
    new ApiResponse(fs)
}

response.failure { FromServer fs, ApiResponse apiResponse ->
    apiResponse
}
```

## Observations

1. It's interesting to note that `ApacheFromServer.hasBody` returns `true` even though `Content-Length` is `0`.
2. I can work around the problem by adding a `response.when(CREATED)` and constructing an empty `ApiResponse` with just the headers and status but since pretty much every HTTP status can be accompanied by an empty body, I don't want to end up having a `response.when` for every HTTP status that we support.

**Comment by cjstehno** [comment]
I am not really understanding what the issue is or the desired result... can you provide an example of a failing unit test for this (even just some pseudo code)?

**Comment by radarsh** [comment]
I can see why you are confused. Give me some time and I will come up with a better explanation.

**Comment by radarsh** [comment]
> Is this just that you feel `hasBody()` should return false when there is no content?

That is probably the root cause of the issue. Anyway what I was trying to explain was that I needed to construct an object that represents the response of an API call (`ApiResponse`) regardless of whether the HTTP response had a body or not. I found that `response.parser` wasn't getting invoked when there was no content. It's perfectly logical for it not to do so.

The work around I had to put in was to have a `response.when` for every status code for which I anticipate an empty body.

I hope I made sense this time :-)

Feel free to close this issue as I feel it's a rather strange requirement.

**Comment by cjstehno** [comment]
This is done in the `development` branch and will be in the 1.0.0 release.
